File: tests/python/test_conflict_key.py
import pytest
from starling.bus.conflict_key import canonical_conflict_key_hex

# Locked parity fixture -- IMPLEMENTER: paste hex from C++ ParityFixtureHex stdout.
PARITY_HEX = "128e262474462a27c39126dbfc4c3876cac63f6d11f53a0161a8b6c8b66f8790"


class ParityStmt:
    holder_id             = "holder-uuid-parity"
    holder_tenant_id      = "tenant-parity"
    subject_kind          = "entity"
    subject_id            = "entity-uuid-parity"
    predicate             = "responsible_for"
    object_kind           = "str"
    object_value          = "auth"
    canonical_object_hash = "aaaa1111bbbb2222cccc3333dddd4444eeee5555ffff6666aaaa1111bbbb2222"
    modality              = "BELIEVES"
    polarity              = "POS"
    confidence            = 0.9
    observed_at           = "2026-01-01T00:00:00Z"
    valid_from            = None
    valid_to              = None
    event_time_start      = None


def test_parity_hex_matches_cpp():
    assert PARITY_HEX != "REPLACE_WITH_CPP_OUTPUT", \
        "Implementer must run C++ test and paste PARITY_HEX value"
    assert canonical_conflict_key_hex(ParityStmt()) == PARITY_HEX


def test_different_holder_produces_different_key():
    class S(ParityStmt):
        holder_id = "different-holder"
    assert canonical_conflict_key_hex(S()) != canonical_conflict_key_hex(ParityStmt())


def test_different_modality_produces_different_key():
    class S(ParityStmt):
        modality = "KNOWS"
    assert canonical_conflict_key_hex(S()) != canonical_conflict_key_hex(ParityStmt())


def test_same_input_produces_same_key():
    assert canonical_conflict_key_hex(ParityStmt()) == canonical_conflict_key_hex(ParityStmt())


def test_interval_participates_in_key():
    class S(ParityStmt):
        valid_from = "2026-01-01T00:00:00Z"
    assert canonical_conflict_key_hex(S()) != canonical_conflict_key_hex(ParityStmt())

File: tests/cpp/test_conflict_key.cpp
#include "starling/bus/conflict_key.hpp"
#include "starling/extractor/extracted_statement.hpp"
#include "starling/schema/statement_enums.hpp"

#include <gtest/gtest.h>
#include <iostream>

namespace {
using namespace starling;
using namespace starling::bus;

extractor::ExtractedStatement make_parity_stmt() {
    extractor::ExtractedStatement s;
    s.holder_id             = "holder-uuid-parity";
    s.holder_tenant_id      = "tenant-parity";
    s.subject_kind          = "entity";
    s.subject_id            = "entity-uuid-parity";
    s.predicate             = "responsible_for";
    s.object_kind           = "str";
    s.object_value          = "auth";
    s.canonical_object_hash = "aaaa1111bbbb2222cccc3333dddd4444eeee5555ffff6666aaaa1111bbbb2222";
    s.modality              = schema::Modality::BELIEVES;
    s.polarity              = schema::Polarity::POS;
    s.confidence            = 0.9;
    s.observed_at           = "2026-01-01T00:00:00Z";
    return s;
}

TEST(ConflictKey, ParityFixtureHex) {
    auto stmt = make_parity_stmt();
    auto hex = canonical_conflict_key_hex(stmt);
    std::cout << "PARITY_HEX=" << hex << std::endl;
    EXPECT_EQ(hex.size(), 64u);
    EXPECT_EQ(hex, "128e262474462a27c39126dbfc4c3876cac63f6d11f53a0161a8b6c8b66f8790");
}

TEST(ConflictKey, DifferentHolderProducesDifferentKey) {
    auto s1 = make_parity_stmt();
    auto s2 = make_parity_stmt();
    s2.holder_id = "different-holder";
    EXPECT_NE(canonical_conflict_key_hex(s1), canonical_conflict_key_hex(s2));
}

TEST(ConflictKey, DifferentModalityProducesDifferentKey) {
    auto s1 = make_parity_stmt();
    auto s2 = make_parity_stmt();
    s2.modality = schema::Modality::KNOWS;
    EXPECT_NE(canonical_conflict_key_hex(s1), canonical_conflict_key_hex(s2));
}

TEST(ConflictKey, DifferentObjectHashProducesDifferentKey) {
    auto s1 = make_parity_stmt();
    auto s2 = make_parity_stmt();
    s2.canonical_object_hash = "0000000000000000000000000000000000000000000000000000000000000000";
    EXPECT_NE(canonical_conflict_key_hex(s1), canonical_conflict_key_hex(s2));
}

TEST(ConflictKey, SameInputProducesSameKey) {
    auto s1 = make_parity_stmt();
    auto s2 = make_parity_stmt();
    EXPECT_EQ(canonical_conflict_key_hex(s1), canonical_conflict_key_hex(s2));
}

TEST(ConflictKey, IntervalParticipatesInKey) {
    auto s1 = make_parity_stmt();
    auto s2 = make_parity_stmt();
    s2.valid_from = "2026-01-01T00:00:00Z";
    EXPECT_NE(canonical_conflict_key_hex(s1), canonical_conflict_key_hex(s2));
}

}  // namespace
